source/extensions/filters/http/aws_lambda/sts_connection_pool.cc

```diff
@@ -0,0 +1,221 @@
+      }
+    }
+
+  private:
+    StsConnectionPool::Context::Callbacks *callbacks_;
+    StsConnectionPoolImpl &parent_;
+    Event::Dispatcher &dispatcher_;
+  };
+
+  using ContextImplPtr = std::unique_ptr<ContextImpl>;
+
+  StsFetcherPtr fetcher_;
+  Api::Api &api_;
+  Event::Dispatcher &dispatcher_;
+  std::string role_arn_;
+  StsConnectionPool::Callbacks *callbacks_;
+
+  std::list<ContextImplPtr> connection_list_;
+
+  bool request_in_flight_ = false;
+};
+
+// using ContextImplPtr = std::unique_ptr<StsConnectionPoolImpl::ContextImpl>;
+
+StsConnectionPoolImpl::StsConnectionPoolImpl(
+    Api::Api &api, Event::Dispatcher &dispatcher,
+    const absl::string_view role_arn, StsConnectionPool::Callbacks *callbacks,
+    StsFetcherPtr fetcher)
+    : fetcher_(std::move(fetcher)), api_(api), dispatcher_(dispatcher),
+      role_arn_(role_arn), callbacks_(callbacks){};
+
+StsConnectionPoolImpl::~StsConnectionPoolImpl() {
+  // When the conn pool is being destructed, make sure to inform all of the
+  // contexts
+  for (auto &&ctx : connection_list_) {
+    ctx->callbacks()->onFailure(CredentialsFailureStatus::ContextCancelled);
+  }
+  // Cancel fetch
+  if (fetcher_ != nullptr) {
+    fetcher_->cancel();
+  }
+};
+
+void StsConnectionPoolImpl::init(const envoy::config::core::v3::HttpUri &uri,
+                                 const absl::string_view web_token) {
+  request_in_flight_ = true;
+  fetcher_->fetch(uri, role_arn_, web_token, this);
+}
+
+StsConnectionPool::Context *
+StsConnectionPoolImpl::add(StsConnectionPool::Context::Callbacks *callbacks) {
+  ContextImpl *ctx_ptr{new ContextImpl(callbacks, *this, dispatcher_)};
+  std::unique_ptr<ContextImpl> ctx{ctx_ptr};
+
+  LinkedList::moveIntoList(std::move(ctx), connection_list_);
+  return ctx_ptr;
+};
+
+void StsConnectionPoolImpl::onSuccess(const absl::string_view body) {
+  ASSERT(body != nullptr);
+  request_in_flight_ = false;
+
+// using a macro as we need to return on error
+// TODO(yuval-k): we can use string_view instead of string when we upgrade to
+// newer absl.
+#define GET_PARAM(X)                                                           \
+  std::string X;                                                               \
+  {                                                                            \
+    std::match_results<absl::string_view::const_iterator> matched;             \
+    bool result = std::regex_search(body.begin(), body.end(), matched,         \
+                                    StsResponseRegex::get().regex_##X);        \
+    if (!result || !(matched.size() != 1)) {                                   \
+      ENVOY_LOG(trace, "response body did not contain " #X);                   \
+      for (auto &&ctx : connection_list_) {                                    \
+        ctx->callbacks()->onFailure(CredentialsFailureStatus::InvalidSts);     \
+      }                                                                        \
```

Comment:
extract this to a method, as this logic is used in dtor too?

we should also empty `(while loop/back()/pop_back())` this list as this may be called again in dtor and onFailure, making the callback invoked twice, similar to how you do it in the bottom of the function.